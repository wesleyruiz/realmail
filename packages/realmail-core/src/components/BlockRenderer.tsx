import {
  AdvancedType,
  BasicType,
  HIDE_DESKTOP_BLOCK_CLASS_NAME,
  HIDE_DESKTOP_INLINE_BLOCK_CLASS_NAME,
  HIDE_MOBILE_BLOCK_CLASS_NAME,
  HIDE_MOBILE_INLINE_BLOCK_CLASS_NAME,
} from '@core/constants';
import { IBlock } from '@core/typings';
import { BlockManager } from '@core/utils';
import { classnames } from '@core/utils/classnames';
import { useEmailRenderContext } from '@core/utils/JsonToMjml';
import { get, set } from 'lodash';
import React from 'react';

type BlockDataItem = Omit<
  Parameters<IBlock['render']>[0],
  'mode' | 'context' | 'dataSource'
>;

const inlineBlockTypes = [
  AdvancedType.GROUP,
  AdvancedType.COLUMN,
  BasicType.GROUP,
  BasicType.COLUMN,
] as string[];

export const BlockRenderer = (props: BlockDataItem) => {
  const { data } = props;
  const { mode, context, dataSource, displayMode } = useEmailRenderContext();
  const block = BlockManager.getBlockByType(data.type);
  if (!block) return null;
  const cloneData = {
    ...data,
    attributes: { ...data.attributes },
  };
  const hideDesktopClassName = inlineBlockTypes.includes(block.type)
    ? HIDE_DESKTOP_INLINE_BLOCK_CLASS_NAME
    : HIDE_DESKTOP_BLOCK_CLASS_NAME;

  const hideMobileClassName = inlineBlockTypes.includes(block.type)
    ? HIDE_MOBILE_INLINE_BLOCK_CLASS_NAME
    : HIDE_MOBILE_BLOCK_CLASS_NAME;

  const cssClass: string = get(cloneData, 'attributes.css-class') || '';

  if (displayMode === 'only-desktop') {
    if (cssClass.includes(hideDesktopClassName)) {
      return null;
    }
    if (!cssClass.includes(hideMobileClassName)) {
      set(cloneData, 'attributes.css-class', classnames(cssClass, hideMobileClassName));
    }
  } else if (displayMode === 'only-mobile') {
    if (cssClass.includes(hideMobileClassName)) {
      return null;
    }
    if (!cssClass.includes(hideDesktopClassName)) {
      set(cloneData, 'attributes.css-class', classnames(cssClass, hideDesktopClassName));
    }
  }
  return <>{block.render({ ...props, data: cloneData, mode, context, dataSource })}</>;
};

// const BlockEditRenderer = (props: BlockDataItem) => {
//   const [refEle, setRefEle] = useState<HTMLElement | null>(null);
//   const { data, renderPortal, ...rest } = props;
//   const block = BlockManager.getBlockByType(data.type);

//   const onCallbackBlockNode = (node: HTMLElement) => {

//     if (!node) return;
//     if (node instanceof HTMLElement) {
//       if (node.classList.contains(getNodeTypeClassName(data.type))) {
//         setRefEle(node);
//       } else {
//         const ele = node.querySelector(`.${getNodeTypeClassName(data.type)}`) as HTMLElement;
//         setRefEle(ele);
//       }

//     }

//   };

//   if (!block) return null;
//   const reactBlock = block.render(props);
//   if (!reactBlock) return null;
//   return (
//     <>
//       {
//         createElement(reactBlock.type, {
//           ...reactBlock.props,
//           ref: onCallbackBlockNode,
//         })
//       }
//       {
//         refEle && renderPortal && createPortal(<>{renderPortal({ ...rest, data, refEle })}</>, refEle)
//       }
//     </>
//   );
// };
