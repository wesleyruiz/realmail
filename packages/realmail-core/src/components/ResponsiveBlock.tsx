import React from 'react';
import { Raw } from './Raw';

import { EmailRenderProvider } from '@core/utils/JsonToMjml';
import { IPage } from '@core/blocks';

export interface ResponsiveBlockProps {
  desktop: React.FC;
  mobile: React.FC;
  mode: 'testing' | 'production';
  dataSource?: { [key: string]: any };
  context?: IPage;
}

export function ResponsiveBlock(props: ResponsiveBlockProps) {
  const { desktop: Desktop, mobile: Mobile } = props;

  return (
    <>
      {/* only visible in desktop */}
      <EmailRenderProvider {...props} displayMode={'only-desktop'}>
        <>
          <Desktop />
        </>
      </EmailRenderProvider>

      {/* only visible in mobile */}
      <Raw>{`<!-- htmlmin:ignore --><!--[if !mso]><!--><!-- htmlmin:ignore -->`}</Raw>
      <EmailRenderProvider {...props} displayMode={'only-mobile'}>
        <>
          <Mobile />
        </>
      </EmailRenderProvider>
      <Raw>{`<!-- htmlmin:ignore --><!--<![endif]--><!-- htmlmin:ignore -->`}</Raw>
    </>
  );
}
