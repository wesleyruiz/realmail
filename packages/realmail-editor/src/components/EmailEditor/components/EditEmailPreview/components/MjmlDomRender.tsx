import React, { useEffect, useMemo, useState } from 'react';
import mjml from 'mjml-browser';
import { getPageIdx, IPage, JsonToMjml } from 'realmail-core';
import { cloneDeep, isEqual } from 'lodash';
import { useEditorContext } from '@/hooks/useEditorContext';
import { HtmlStringToReactNodes } from '@/utils/HtmlStringToReactNodes';
import { createPortal } from 'react-dom';
import { useEditorProps } from '@/hooks/useEditorProps';
import { getEditorRoot, getShadowRoot } from '@/utils';
import { DATA_RENDER_COUNT, FIXED_CONTAINER_ID } from '@/constants';
import { useValidationContext } from '@/components/Provider/EmailEditorProvider';
import { useActiveTab } from '@/hooks/useActiveTab';
import { ActiveTabKeys } from '@/components/Provider/BlocksProvider';

let count = 0;
export function MjmlDomRender() {
  const { pageData: content } = useEditorContext();
  const [pageData, setPageData] = useState<IPage | null>(null);
  const [ref, setRef] = useState<HTMLDivElement | null>(null);
  const { dashed, mergeTags, enabledMergeTagsBadge } = useEditorProps();
  const [isTextFocus, setIsTextFocus] = useState(false);
  const { errorBlocksMap } = useValidationContext();
  const { activeTab } = useActiveTab();

  const isTextFocusing = getShadowRoot().activeElement?.getAttribute('contenteditable') === 'true';

  useEffect(() => {
    if (!isTextFocus && !isEqual(content, pageData)) {

      setPageData(cloneDeep(content));
    }
  }, [content, pageData, setPageData, isTextFocus]);

  useEffect(() => {
    setIsTextFocus(isTextFocusing);
  }, [isTextFocusing]);

  useEffect(() => {
    const onClick = (e: MouseEvent) => {

      if (getShadowRoot()?.contains(e.target as Node)) {
        return;
      }

      const fixedContainer = document.getElementById(FIXED_CONTAINER_ID);
      if (fixedContainer?.contains(e.target as Node)) {
        return;
      }
      // element has unmount
      if (!document.body.contains(e.target as Node)) {
        return;
      }
      setIsTextFocus(false);
    };

    window.addEventListener('click', onClick);
    return () => {
      window.removeEventListener('click', onClick);
    };
  }, []);

  useEffect(() => {
    const root = getShadowRoot();
    if (!root) return;
    const onClick = (e: Event) => {
      const isFocusing =
        getShadowRoot().activeElement?.getAttribute('contenteditable') ===
        'true';
      if (isFocusing) {
        setIsTextFocus(true);
      }
    };

    root.addEventListener('click', onClick);
    return () => {
      root.removeEventListener('click', onClick);
    };
  }, []);

  const html = useMemo(() => {
    if (!pageData) return '';
    const cloneData = cloneDeep(pageData);
    if (activeTab === ActiveTabKeys.MOBILE) {
      cloneData.data.value.breakpoint = '2000px'; // 强行模拟移动端行为
    }
    const renderHtml = mjml(
      JsonToMjml({
        data: cloneData,
        idx: getPageIdx(),
        context: cloneData,
        mode: 'testing',
        dataSource: cloneDeep(mergeTags),
      })
    ).html;
    return renderHtml;
  }, [mergeTags, pageData, activeTab]);

  return useMemo(() => {
    return (
      <div
        {
        ...{
          [DATA_RENDER_COUNT]: count++
        }
        }
        data-dashed={dashed}
        ref={setRef}
        style={{
          outline: 'none',
          position: 'relative',
        }}
        role='tabpanel'
        tabIndex={0}
      >
        {ref &&
          createPortal(
            HtmlStringToReactNodes(html, {
              enabledMergeTagsBadge: Boolean(enabledMergeTagsBadge),
            }),
            ref
          )}
        <style>
          {
            Object.keys(errorBlocksMap).map(idx => {
              return `
                [data-idx="${idx}"] {
                  outline: 2px solid var(--error-color);
                }
            `;
            }).join('\n')

          }
        </style>
      </div>
    );
  }, [dashed, ref, html, enabledMergeTagsBadge, errorBlocksMap]);
};
