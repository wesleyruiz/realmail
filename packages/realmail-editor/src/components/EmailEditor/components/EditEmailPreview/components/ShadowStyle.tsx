import React from 'react';
import iconfontText from '@/assets/font/iconfont.css?inline';
import styles from '@/styles/block-shadowDom-interactive.scss?inline';
import { useEditorProps } from '@/hooks/useEditorProps';

export function ShadowStyle() {
  const style = getComputedStyle(document.body);

  const {
    interactiveStyle: {
      hoverColor = `rgb(${style.getPropertyValue('--primary-4') || '24,144,255'})`,
      selectedColor = `rgb(${style.getPropertyValue('--primary-6') || '24,144,255'})`,
      errorColor = 'transparent',
    } = {},
  } = useEditorProps();

  return (
    <>
      <style>{iconfontText}</style>

      <style
        dangerouslySetInnerHTML={{
          __html: `
            * {
              --hover-color: ${hoverColor};
              --selected-color: ${selectedColor};
              --error-color: ${errorColor};
            }

            :host(*){
              all: initial;
            }

            html {
              overflow: overlay !important;
            }

            ${styles}

            `,
        }}
      />
    </>
  );
}
