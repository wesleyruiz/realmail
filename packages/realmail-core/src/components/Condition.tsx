import React from 'react';
import { MjmlBlock } from '@core/components';
import { isUndefined } from 'lodash';
import { ICondition } from '@core/blocks/logic/Condition';
import { LogicType } from '@core/constants';

export type ConditionProps = ICondition['data']['value'] & {
  truthy: React.ReactElement;
  falsy?: React.ReactElement;
  showTruthyInEdit?: boolean;
};

export function Condition(props: ConditionProps) {
  const { expression, showTruthyInEdit } = props;
  return (
    <MjmlBlock
      value={{
        expression,
        showTruthyInEdit,
      }}
      type={LogicType.CONDITION}
    >
      {props.truthy}
      {isUndefined(props.falsy) ? null : props.falsy}
    </MjmlBlock>
  );
}
