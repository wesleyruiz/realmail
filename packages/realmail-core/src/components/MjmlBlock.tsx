import React, { useMemo } from 'react';
import { BlockManager } from '@core/utils';
import { IBlockData, RecursivePartial } from '@core/typings';
import { isString, set } from 'lodash';
import { useEmailRenderContext } from '@core/utils/JsonToMjml';
import { BlockRenderer } from './BlockRenderer';

export interface MjmlBlockProps<T extends IBlockData> {
  idx?: string | null;
  type: T['type'];
  value?: RecursivePartial<T['data']['value']>;
  attributes?: RecursivePartial<T['attributes']>;
  children?: React.ReactNode;
}

export default function MjmlBlock<T extends IBlockData>({
  idx,
  value,
  type,
  attributes,
  children,
}: MjmlBlockProps<T>) {
  const { mode } = useEmailRenderContext();
  const block = BlockManager.getBlockByType(type);

  if (!mode) {
    throw new Error('mode is required!');
  }

  if (!block) {
    throw new Error(`Can no find ${type}`);
  }

  const mergeValue = useMemo((): undefined | {} => {
    if (!value) {
      value = {} as Record<string, any>;
    }

    if (typeof children === 'string') {
      set(value, 'content', children);
    } else if (Array.isArray(children) && children.every(isString)) {
      set(value, 'content', children.join(''));
    }

    return value;
  }, [children, value]);

  return (
    <BlockRenderer
      idx={idx}
      data={{
        type: block.type,
        data: {
          value: mergeValue,
        },
        attributes,
        children: [],
      }}
      children={children}
    />
  );
}
