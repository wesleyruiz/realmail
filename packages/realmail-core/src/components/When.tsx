import React from 'react';
import { MjmlBlock } from '@core/components';
import { ICondition } from '@core/blocks/logic/Condition';
import { LogicType } from '@core/constants';

export type ConditionProps = ICondition['data']['value'] & {
  children: React.ReactElement;
  showTruthyInEdit?: boolean;
};

export function When(props: ConditionProps) {
  const { expression, showTruthyInEdit } = props;
  return (
    <MjmlBlock
      value={{
        expression,
        showTruthyInEdit,
      }}
      type={LogicType.CONDITION}
    >
      {props.children}
      <></>
    </MjmlBlock>
  );
}
