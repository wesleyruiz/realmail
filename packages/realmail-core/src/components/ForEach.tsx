import { MjmlBlock } from '@core/components';
import { IForEach } from '@core/blocks/logic/ForEach';
import React from 'react';
import { MjmlBlockProps } from './MjmlBlock';
import { LogicType } from '@core/constants';

export type ForEachProps = IForEach['data']['value'] & {
  children?: MjmlBlockProps<IForEach>['children'];
};

export function ForEach(props: ForEachProps) {
  const { children, ...rest } = props;

  return (
    <MjmlBlock
      value={{
        ...rest,
      }}
      type={LogicType.FOR_EACH}
    >
      {children}
    </MjmlBlock>
  );
}
