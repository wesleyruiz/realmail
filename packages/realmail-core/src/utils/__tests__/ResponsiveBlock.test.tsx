import { Column, ResponsiveBlock, Text } from '@core/components';
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { unescape } from 'lodash';

describe('Test parseXml', () => {
  const desktop = <Text font-size="24px">111</Text>;
  const mobile = <Text font-size="12px">111</Text>;
  const com = unescape(
    renderToStaticMarkup(
      <ResponsiveBlock
        mode="production"
        blockData={{
          type: 'text',
          data: {
            value: {
              content: '111',
            },
          },
          attributes: {
            padding: '0px 25px 0px 25px',
            align: 'center',
          },
          children: [],
        }}
        desktop={desktop}
        mobile={mobile}
      />,
    ),
  );
  it('render as expect', () => {
    expect(com).toMatchSnapshot();
  });
});

describe('Test parseXml nested blocks', () => {
  const desktop = (
    <Column>
      <Text font-size="24px">111</Text>
    </Column>
  );
  const mobile = (
    <Column>
      <Text font-size="12">111</Text>
    </Column>
  );
  const com = unescape(
    renderToStaticMarkup(
      <ResponsiveBlock
        mode="production"
        blockData={{
          type: 'column',
          data: {
            value: {
              content: '111',
            },
          },
          attributes: {
            padding: '0px 25px 0px 25px',
            align: 'center',
          },
          children: [],
        }}
        desktop={desktop}
        mobile={mobile}
      />,
    ),
  );
  it('render as expect', () => {
    expect(com).toMatchSnapshot();
  });
});
