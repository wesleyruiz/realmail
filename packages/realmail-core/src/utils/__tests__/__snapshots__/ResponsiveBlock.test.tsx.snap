// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test parseXml nested blocks render as expect 1`] = `"<mj-column css-class=\\"hide-mobile-block\\" ><mj-text font-size=\\"24px\\" >111</mj-text></mj-column><mj-raw ><!-- htmlmin:ignore --><!--[if !mso]><!--><!-- htmlmin:ignore --></mj-raw><mj-column css-class=\\"hide-mobile-block\\" ><mj-text font-size=\\"12\\" >111</mj-text></mj-column><mj-raw ><!-- htmlmin:ignore --><!--<![endif]--><!-- htmlmin:ignore --></mj-raw>"`;

exports[`Test parseXml render as expect 1`] = `"<mj-text font-size=\\"24px\\" css-class=\\"hide-mobile-block\\" >111</mj-text><mj-raw ><!-- htmlmin:ignore --><!--[if !mso]><!--><!-- htmlmin:ignore --></mj-raw><mj-text font-size=\\"12px\\" css-class=\\"hide-mobile-block\\" >111</mj-text><mj-raw ><!-- htmlmin:ignore --><!--<![endif]--><!-- htmlmin:ignore --></mj-raw>"`;
